### PR TITLE
Run distilGPT2 on aihwkit

### PR DESCRIPTION
**Pull request by anchen25**
## Description

Implementing GPT-2 based model distilgpt2 on the wikitext-2-raw-v1 dataset using AIHWKit. The example demonstrates how to convert the model to analog, run fine-tuning, text-generation, and inference. 

## Details

1. Model and Dataset:

- Implemented an example using the smallest GPT-2 model (distilgpt2).
- Utilized the wikitext-2-raw-v1 dataset for training and validation, which is smaller and faster to process compared to
- openwebtext.

2. Training and Inference Setup:

- Configured the model to use analog inference with specified noise levels.
- Added support for digital inference as an option.
- Implemented preprocessing functions to handle dataset tokenization.
- Provided functionality to train the model and save/load checkpoints.

3. Logging and Monitoring:

- Integrated TensorBoard for logging training and validation metrics.
- Added TensorBoardCallback to the Trainer for seamless logging.
- Configured the script to save logs in a specific directory and visualize them using TensorBoard.

4. Performance Metrics:

- Calculated validation loss and perplexity as the primary performance metrics.
- Digital model loss varies with learning rate. Lowest training loss = 3.26, lowest inference loss = 3.55
- Loss of HWA-finetuned analog model depends on learning rate and optimizer. Lowest HWA inference loss = 3.55, matching digital model.  

5. How to use command-line arguments:

**_For text generation (of both digital and analog models), use command line arguments: "gt", "L", "c", "pt"_**

Example 1: python 36_gpt2_on_wikitext_v3.py -gt -pt "Once upon a time"    --->   text generation using the  pre-trained DistilGPT2 model without fine-tuning 
Example 2: python 36_gpt2_on_wikitext_v3.py -gt -pt "Once upon a time" -L -c "checkpoint_filename.pth"   --->   Text generation using a fine-tuned (digital or analog) model with a saved checkpoint file

**_For digital model fine-tuning and loss calculation, use command line arguments: "d", "c", "lr", "L"_**   

Example 3: python 36_gpt2_on_wikitext_v3.py -d -lr 1e-5 -c "checkpoint_filename.pth"   --->   fine-tune the digital model with specified learning-rate and save checkpoint to the specified fine name 
Example 4: python 36_gpt2_on_wikitext_v3.py -d -L   --->   inference (loss calculation) without fine-tuning on the pre-trained DistilGPT2 model 
Example 5: python 36_gpt2_on_wikitext_v3.py -d -L -c "checkpoint_filename.pth"   --->   inference (loss calculation) on a fine-tuned digital model with a saved checkpoint file  

**_For analog model HWA fine-tuning and loss calculation, use command line arguments: "t", "c", "n", "lr", "L"_**  

Example 6: python 36_gpt2_on_wikitext_v3.py -t -n 0.0 -l 0.01 -c "checkpoint_filename.pth"   ---> fine-tune the analog model with specified noise and learning rate, and save the checkpoint file 
Example 7: python 36_gpt2_on_wikitext_v3.py -L -c "checkpoint_filename.pth"   --->   inference (loss calculation) on a fine-tuned analog model with a saved checkpoint file

### CODE REVIEW

**Comment by PabloCarmona** [comment]
Is this PR the most recent and that will be used instead of this one: https://github.com/IBM/aihwkit/pull/664? In that case please let us know @anchen25 @charles-mackin  to close the other in favor of this one. Thanks!

File: examples/36_gpt2_on_wikitext_v3.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        print("Error: train_dataset or eval_dataset is None.")
+        return
+    # Create optimizer and trainer
+    optimizer = create_optimizer(model, ARGS.digital)     # type: ignore[no-untyped-call]
+    trainer, writer = make_trainer(model, optimizer,  # type: ignore[no-untyped-call]
+                                   train_dataset, eval_dataset)  # type: ignore[no-untyped-call]
+    # If "-L", load checkpoint file
+    if ARGS.load and ARGS.checkpoint is not None:
+        print(f"Load model from '{ARGS.checkpoint}'.")
+        model.load_state_dict(torch_load(ARGS.checkpoint), strict=False)
```
Comment (charlesmackin) [inline examples/36_gpt2_on_wikitext_v3.py]: Modify to: `Model.load_state_dict(torch_load(ARGS.checkpoint, weights_only=False), strict=False)
`
This will avoid the following warning: 

(1) In PyTorch 2.6, we changed the default value of the `weights_only` argument in `torch.load` from `False` to `True`. Re-running `torch.load` with `weights_only` set to `False` will likely succeed, but it can result in arbitrary code execution. Do it only if you got the file from a trusted source.
(2) Alternatively, to load with `weights_only=True` please check the recommended steps in the following error message.

File: examples/36_gpt2_on_wikitext_v3.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    # Finetune digital or analog model
+    if (ARGS.train_hwa or ARGS.digital) and not ARGS.load:
+        trainer.train()
+        torch_save(model.state_dict(), ARGS.checkpoint)
+    # Calculate inference loss
+    do_inference(model, trainer, writer, ARGS.digital)  # type: ignore[no-untyped-call]
+
+
+def generate_text(prompt, model, max_length=50) -> str:   # type: ignore[no-untyped-def]
```
Comment (charlesmackin) [inline examples/36_gpt2_on_wikitext_v3.py]: Got the following warning:

> The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.

Generally ok, but I would suggest changing generate_text to the following to avoid this warning/issue:

```
def generate_text(prompt, model, max_length=50) -> str:   # type: ignore[no-untyped-def]
    """ Generate a text from the model using a prompt """

    device = next(model.parameters()).device

    encoding = TOKENIZER(prompt, return_tensors='pt')
    input_ids = encoding.input_ids.to(device)
    attention_mask = encoding.attention_mask.to(device)

    outputs = model.generate(
        input_ids,
        attention_mask=attention_mask,
        max_length=max_length,
        num_return_sequences=1,
        do_sample=True,
        temperature=0.7,
        top_k=50,
        repetition_penalty=5.0,
        pad_token_id=TOKENIZER.eos_token_id,
    )
    text = TOKENIZER.decode(outputs[0], skip_special_tokens=True)
    return text
```
